src/main/java/net/mcreator/workspace/references/ReferencesFinder.java

```diff
@@ -0,0 +1,309 @@
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+package net.mcreator.workspace.references;
+
+import net.mcreator.blockly.data.BlocklyXML;
+import net.mcreator.element.GeneratableElement;
+import net.mcreator.element.parts.Sound;
+import net.mcreator.element.parts.procedure.Procedure;
+import net.mcreator.generator.GeneratorWrapper;
+import net.mcreator.generator.mapping.MappableElement;
+import net.mcreator.ui.workspace.resources.TextureType;
+import net.mcreator.util.FilenameUtilsPatched;
+import net.mcreator.workspace.IWorkspaceProvider;
+import net.mcreator.workspace.Workspace;
+import net.mcreator.workspace.elements.ModElement;
+import net.mcreator.workspace.elements.SoundElement;
+import net.mcreator.workspace.resources.Model;
+import net.mcreator.workspace.resources.TexturedModel;
+
+import javax.annotation.Nullable;
+import java.io.File;
+import java.lang.reflect.*;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+import java.util.function.BiPredicate;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+
+public class ReferencesFinder {
+
+	public static Set<ModElement> searchModElementUsages(Workspace workspace, ModElement element) {
+		Set<ModElement> elements = new HashSet<>();
+
+		String query = element.getName();
+		workspace.getModElements().stream().filter(me -> !me.equals(element)).forEach(me -> {
+			GeneratableElement ge = me.getGeneratableElement();
+			if (anyValueMatches(ge, String.class, e -> e.isAnnotationPresent(ModElementReference.class), (a, t) -> {
+				ModElementReference ref = a.getAnnotation(ModElementReference.class);
+				return !Set.of(ref.defaultValues()).contains(t) && query.equals(
+						GeneratorWrapper.getElementPlainName(t));
+			})) {
+				elements.add(me);
+			} else if (anyValueMatches(ge, MappableElement.class, e -> e.isAnnotationPresent(ModElementReference.class),
+					(a, t) -> query.equals(GeneratorWrapper.getElementPlainName(t.getUnmappedValue())))) {
+				elements.add(me);
+			} else if (anyValueMatches(ge, Procedure.class, e -> e.isAnnotationPresent(ModElementReference.class),
+					(a, t) -> t.getName() != null && !t.getName().isEmpty() && !t.getName().equals("null")
+							&& query.equals(t.getName()))) {
+				elements.add(me);
+			} else if (anyValueMatches(ge, String.class, e -> e.isAnnotationPresent(BlocklyXML.class),
+					(a, t) -> t.contains(">CUSTOM:" + query + "</field>") || t.contains(">" + query + "</field>"))) {

```

Comment:
Could you provide example of case in procedure where ME is referenced without CUSTOM: so the other check is also needed?

Btw I have slightly modified this method in https://github.com/MCreator/MCreator/pull/3500/commits/30d33e06d2071bbe00e55c3da95b89ba1a0f3178 so CUSTOM: is not added to the query and then also added to the "t" ME plain name again as this is unneeded performance burden. Please review if I missed something when refactoring. Thanks!

Reply:
In cases where all possible elements are custom (e.g. procedures), the prefix is not needed, so here is this second check.

Reply:
<details><summary>Here's an example showing all types of references: external only, external and custom, custom only.</summary><p>

```xml
<block type="event_trigger" deletable="false" x="40" y="40">
	<field name="trigger">no_ext_trigger</field>
	<next><block type="controls_if">
		<value name="IF0"><block type="logic_binary_ops" inline="false">
			<field name="OP">XOR</field>
			<value name="A"><block type="itemstack_has_enchantment">
				<field name="enhancement">FLAME</field>
				<value name="item"><block type="mcitem_all">
					<field name="value">Blocks.CHERRY_SAPLING</field>
				</block></value>
			</block></value>
			<value name="B"><block type="item_istype">
				<field name="item_type">Block</field>
				<value name="item"><block type="mcitem_all">
					<field name="value">CUSTOM:Testive</field>
				</block></value>
			</block></value>
		</block></value>
		<statement name="DO0"><block type="call_procedure">
			<field name="procedure">ExProc</field>
		</block></statement>
	</block></next>
</block>
```
</p></details>

![image](https://github.com/MCreator/MCreator/assets/71347607/e24fae4b-13da-45f0-b0fb-0a17cb3fc5a5)

Reply:
I see, thank you!